File: LeetCode/Study Plan/Practice Problems/Medium/Print Words Vertivally/Go/print_words_vertically.go
package main

import (
	"fmt"
	"strings"
)

func print_words_vertically(s string) []string {
	res := []string{}
	str := strings.Split(s, " ")
	n := len(str)
	m := len(str[0])
	for i := 0; i < n; i++ {
		m = max(m, len(str[i]))
	}
	for j := 0; j < m; j++ {
		s := ""
		for i := 0; i < n; i++ {
			if len(str[i]) <= j {
				s += " "
				continue
			}
			s += string(str[i][j])
		}
		s = strings.TrimRight(s, " ")
		res = append(res, s)
	}
	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	fmt.Println(print_words_vertically("HOW ARE YOU"))
}

File: LeetCode/Study Plan/Practice Problems/Medium/Print Words Vertivally/Cpp/print_words_vertically.cc
#include <iostream>
#include <vector>
#include <sstream>

#define PB push_back 

typedef std::vector<std::string> vs;

vs split(std::string s, char delim, int &len) {
    vs st;
    std::stringstream ss(s);
    std::string token;
    while(getline(ss, token, delim)) {
        len = std::max(len, (int)token.size());
        st.PB(token);
    }
    return st;
}

vs print_words_vertically(std::string s) {
    int len = 0;
    vs split_words = split(s, ' ', len);
    vs res;
    for (int i=0;i<len;i++) {
        std::string temp = "";
        for (int j=0; j<split_words.size(); j++) {
            if (split_words[j].size() <= i) {
                temp += " ";
                continue;
            }
            temp += split_words[j][i];
        }
        int k = 0;
        for (k=temp.size()-1; k>=0; k--) {
            if (temp[k] != ' ') break;
        }
        std::string t = "";
        for (int j=0;j<=k;j++) {
            t += temp[j];
        }
        res.PB(temp);
    }
    return res;
}

int main() {
    vs res = print_words_vertically("HOW ARE YOU");
    for (auto it=res.begin(); it!= res.end(); it++) std::cout<<*it<<"\n";
}
